File: root/Worker.py
from __future__ import annotations
from typing import Optional, TYPE_CHECKING

if TYPE_CHECKING:
    from root.CashDesk import CashDesk
    from root.CoffeeHouse import CoffeeHouse
    from root.Product import Product
    from root.Recipe import Recipe
    from root.RecipeStep import RecipeStep


class Worker:
    _cashDesk: CashDesk
    _model: CoffeeHouse
    _currentProduct: Optional[Product]
    _currentRecipe: Optional[Recipe]
    _recipeStepIndex: int
    _nextActionTime: int

    def __init__(self, model: CoffeeHouse, cashDesk: CashDesk):
        self._currentRecipe = None
        self._currentProduct = None
        self._recipeStepIndex = 0
        self._nextActionTime = 0
        self._model = model
        self._cashDesk = cashDesk

    def process(self, currentTime: int) -> None:
        if self._currentRecipe is None:
            visitor = self._cashDesk.getFirstVisitor()
            if visitor is None:
                return None
            self._currentProduct = visitor.getDesiredProduct()
            if self._currentProduct.isNeedPrepare():
                self._currentRecipe = self._model.getRecipe(self._currentProduct)
                print(f'Time={currentTime} Worker({self.__hash__()}) start prepare product {self._currentProduct}')
            else:
                visitor.askDesiredProduct(currentTime)
        if self._currentRecipe is not None:
            self._prepareProductProcess(currentTime)

    def _prepareProductProcess(self, currentTime: int) -> None:
        if self._nextActionTime == 0:
            self._recipeStepIndex = 0
            step = self.__getRecipeStep()
            assert step is not None

            self._nextActionTime = currentTime
            self._processStepTime(step)
        elif self._nextActionTime <= currentTime:
            step = self.__getRecipeStep()
            assert step is not None
            result = self._model.takeIngredient(step.getIngredient(), step.getAmount())
            if not result:
                self._cashDesk.getFirstVisitor().failDesiredProduct(currentTime, self._currentProduct)
                self._prepareProductFinished(currentTime)
            self._recipeStepIndex += 1
            nextStep = self.__getRecipeStep()
            if nextStep is None:
                # finish prepare
                self._model.takeProduct(self._currentProduct)
                self._cashDesk.getFirstVisitor().givePreparedProduct(currentTime, self._currentProduct)
                self._prepareProductFinished(currentTime)
            else:
                self._processStepTime(nextStep)
                self._prepareProductProcess(currentTime)

    def _prepareProductFinished(self, currentTime: int) -> None:
        print(f'Time={currentTime} Worker({self.__hash__()}) finish prepare {self._currentProduct}')
        self._currentProduct = None
        self._currentRecipe = None
        self._recipeStepIndex = 0
        self._nextActionTime = 0

    def _processStepTime(self, step: RecipeStep):
        self._nextActionTime += step.getTime()

    def __getRecipeStep(self) -> Optional[RecipeStep]:
        assert self._currentRecipe is not None
        stepList = self._currentRecipe.getStepList()
        if len(stepList) > self._recipeStepIndex:
            return stepList[self._recipeStepIndex]
        else:
            return None

File: cplusplus/Worker.cpp
#include "Worker.hpp"

Worker::Worker(CoffeeHouse *model, CashDesk *cashDesk) : model(model), cashDesk(cashDesk)
{
	this->cashDesk = cashDesk;
	this->model = model;
}

void Worker::process(long currentTime)
{
	if (currentRecipe == nullptr)
	{
		Visitor *visitor = cashDesk->getFirstVisitor();
		if (visitor == nullptr)
		{
			return;
		}

		currentProduct = visitor->getDesiredProduct();
		if (currentProduct->getIsNeedPrepare())
		{
			currentRecipe = model->getRecipe(currentProduct);
			std::cout << "Time=" + std::to_string(currentTime) + " Worker(" + getHashCode() + ") start prepare product " << currentProduct->toString() << std::endl;
		}
		else
		{
			visitor->askDesiredProduct(currentTime);
		}
	}

	if (currentRecipe != nullptr)
	{
		prepareProductProcess(currentTime);
	}
}

void Worker::prepareProductProcess(long currentTime)
{
	if (nextActionTime == 0)
	{
		recipeStepIndex = 0;
		auto step = getRecipeStep(recipeStepIndex);
		assert(step != nullptr);

		nextActionTime = currentTime;
		processStepTime(step);
	}
	else if (nextActionTime <= currentTime)
	{
		auto step = getRecipeStep(recipeStepIndex);
		assert(step != nullptr);

		auto result = model->takeIngredient(step->getIngredient(), step->getAmount());
		if (!result)
		{
			cashDesk->getFirstVisitor()->failDesiredProduct(currentTime, currentProduct);
			prepareProductFinished(currentTime);
		}
		recipeStepIndex += 1;
		auto nextStep = getRecipeStep(recipeStepIndex);
		if (nextStep == nullptr)
		{
			// finish prepare
			model->takeProduct(currentProduct);
			cashDesk->getFirstVisitor()->givePreparedProduct(currentTime, currentProduct);
			prepareProductFinished(currentTime);
		}
		else
		{
			processStepTime(nextStep);
			prepareProductProcess(currentTime);
		}
	}
}

void Worker::prepareProductFinished(long currentTime)
{
	std::cout << "Time=" + std::to_string(currentTime) + " Worker(" + getHashCode() + ") finish prepare " << currentProduct->toString() << std::endl;
	nextActionTime = 0;
	recipeStepIndex = 0;
	currentProduct = nullptr;
	currentRecipe = nullptr;
}

void Worker::processStepTime(RecipeStep *step)
{
	nextActionTime += step->getTime();
}

std::string Worker::getHashCode()
{
	std::string thisAddress = std::to_string(reinterpret_cast<int>(this));
	std::string hashCode = std::to_string(std::hash<std::string>()(thisAddress));
	
	return hashCode;
}

RecipeStep *Worker::getRecipeStep(int index)
{
	assert(currentRecipe != nullptr);
	auto stepVector = currentRecipe->getStepVector();
	return stepVector.size() > recipeStepIndex ? stepVector[recipeStepIndex] : nullptr;
}